src/protocol/ipa_prf/mod.rs

```diff
@@ -1,7 +1,222 @@
 pub mod shuffle;
+
+#[cfg(feature = "descriptive-gate")]
+#[derive(Step)]
+pub(crate) enum Step {
+    ConvertFp25519,
+    EvalPrf,
+}
+
+#[cfg(feature = "descriptive-gate")]
+#[derive(Debug)]
+pub struct PrfIpaInputRow<BK: GaloisField, TV: GaloisField, TS: GaloisField> {
+    pub match_key: Replicated<BA64>,
+    pub is_trigger_bit: Replicated<Gf2>,
+    pub breakdown_key: Replicated<BK>,
+    pub trigger_value: Replicated<TV>,
+    pub timestamp: Replicated<TS>,
+}
+
+/// IPA OPRF Protocol
+///
+/// We return `Replicated<F>` as output.
+/// This protocol does following steps
+/// 1. Shuffles the input (TBD)
+/// 2. Converts boolean arrays of match keys to elliptical values
+/// 3. Computes OPRF on the match keys and reveals the OPRF
+/// 4. Sorts inputs based on reveal oprf and timestamp (TBD)
+/// 5. Computes the attribution, caps results and aggregates
+/// # Errors
+/// Propagates errors from config issues or while running the protocol
+/// # Panics
+/// Propagates errors from config issues or while running the protocol
+#[cfg(feature = "descriptive-gate")]
+pub async fn oprf_ipa<C, BK, TV, TS, F>(
+    ctx: C,
+    input_rows: Vec<PrfIpaInputRow<BK, TV, TS>>,
+    config: IpaQueryConfig,
+) -> Result<Vec<Replicated<F>>, Error>
+where
+    C: UpgradableContext,
+    C::UpgradedContext<Gf2>: UpgradedContext<Gf2, Share = Replicated<Gf2>>,
+    C::UpgradedContext<F>: UpgradedContext<F, Share = Replicated<F>>,
+    BK: GaloisField,
+    TV: GaloisField,
+    TS: GaloisField,
+    F: PrimeField + ExtendableField,
+    Replicated<F>: Serializable,
+{
+    let user_cap: i32 = config.per_user_credit_cap.try_into().unwrap();
+    assert!(
+        user_cap & (user_cap - 1) == 0,
+        "This code only works for a user cap which is a power of 2"
+    );
+
+    // TODO (richaj): Add shuffle either before the protocol starts or, after converting match keys to elliptical curve.
+    // We might want to do it earlier as that's a cleaner code
+
+    let convert_ctx = ctx
+        .narrow(&Step::ConvertFp25519)
+        .set_total_records(input_rows.len());
+    let eval_ctx = ctx
+        .narrow(&Step::EvalPrf)
+        .set_total_records(input_rows.len());
+
+    let prf_key = gen_prf_key(&convert_ctx);
+
+    let pseudonymed_user_ids = ctx
+        .parallel_join(input_rows.iter().zip(repeat(prf_key)).enumerate().map(
+            |(idx, (record, prf_key))| {
+                let convert_ctx = convert_ctx.clone();
+                let eval_ctx = eval_ctx.clone();
+                async move {
+                    let record_id = RecordId::from(idx);
+                    let prf_of_match_key =
+                        convert_to_fp25519::<_, BA64>(convert_ctx, record_id, &record.match_key)
+                            .await?;
+                    eval_dy_prf(eval_ctx, record_id, &prf_key, &prf_of_match_key).await
+                }
+            },
+        ))
+        .await?;
+    let pseudonymed_inputs = input_rows
```

Comment:
this second iteration over the input looks unnecessary. At the end of PRF evaluation circuit, the input row is ready and we could probably yield `PrfShardedIpaInputRow`

Reply:
Agree on this but unable to get past a compile error. https://github.com/richajaindce/raw-ipa/pull/33